etna/lib/etna/filesystem.rb

```diff
@@ -377,55 +377,151 @@ def exist?(src)
+      def url(src)
+        "sftp://#{@host}/#{src}"
+      end
+
+      def authn
+        "#{@username}:#{@password}"
+      end
+
+      def curl_cmd(path, opts=[])
+        connection = Curl::Easy.new(url(path))
+        connection.http_auth_types = :basic
+        connection.username = @username
+        connection.password = @password
+
+        connection
+      end
+
+      def sftp_file_from_path(src)
+        file = ls(::File.dirname(src)).split("\n").map do |listing|
+          SftpFile.new(listing)
+        end.select do |file|
+          file.name == ::File.basename(src)
+        end
+
+        raise "#{src} not found" if file.empty?
+
+        file.first
       end
 
-      def sftp
-        @sftp ||= begin
-          conn = Net::SFTP::Session.new(ssh)
-          conn.loop { conn.opening? }
+      def mkcommand(rd, wd, file, opts, size_hint: nil)
+        env = {}
+        cmd = [env, "curl"]
 
-          conn
+        cmd << "-u"
+        cmd << authn
+        cmd << "-o"
+        cmd << "-"
+        cmd << "-N"
+        cmd << url(file)
+
+        if opts.include?('r')
+          cmd << {out: wd}
         end
+
+        cmd
       end
 
       def with_readable(src, opts = 'r', &block)
-        sftp.file.open(src, opts, &block)
-      end
+        raise "#{src} does not exist" unless exist?(src)
 
-      def ls(dir)
-        sftp.dir.entries(dir)
+        sftp_file = sftp_file_from_path(src)
+
+        mkio(src, opts, size_hint: sftp_file.size, &block)
       end
 
       def exist?(src)
-        begin
-          sftp.file.open(src)
-        rescue Net::SFTP::StatusException
-          return false
-        end
-        return true
+        files = ls(::File.dirname(src))
+        files.include?(::File.basename(src))
+      end
+
+      def ls(dir)
+        dir = dir + "/" unless "/" == dir[-1]  # Trailing / makes curl list directory
+
+        return @dir_listings[dir] if @dir_listings.has_key?(dir)
+
+        listing = ''
+        connection = curl_cmd(dir)
+        connection.on_body { |data| listing << data; data.size }
+        connection.perform
+
+        @dir_listings[dir] = listing
+
+        listing
       end
 
       def stat(src)
-        sftp.file.open(src).stat
+        sftp_file_from_path(src)
+      end
+    end
+
+    class RemoteSSH < Filesystem
+      class RemoteSSHError < Exception
+      end
+
+      def initialize(host:, username:, password: nil, port: 22, root:, **args)
+        @username = username
+        @password = password
+        @host = host
+        @port = port
+        @root = root
+      end
+
+      def ssh
+        @ssh ||= Net::SSH.start(@host, @username, password: @password)
+      end
+
+      def mkdir_p(dir)
+        output = ssh.exec!("mkdir -p #{dir}")
+
+        raise RemoteSSHError.new("Unable to mkdir -p, #{output}") unless 0 == output.exitstatus
+      end
+
+      def lftp_get(username:, password:, host:, remote_filename:, &block)
```

Comment:
So I guess the lftp username/password is not necessarily the same as the ssh one?
Oof.  This one is interesting case.

Reply:
Yeah, this is LFTP from the remote SSH host to a different server (i.e. from Polyphemus, SSH to C4, then from C4 use LFTP to fetch the file from CAT onto C4 -- so different credentials for both C4 and the CAT).

Reply:
It does feel like the `Filesystem` abstraction may not be the right fit for this use case, so not sure if I should try to stream the file from CAT -> Polyphemus -> C4 (seems like an unnecessary network hop / bottleneck) like the other `Filesystem` impls (which seem to make more sense when one endpoint is Metis), or implement this as something different than a `Filesystem`. Thoughts, @corps ?